### ISSUE TITLE
Cannot signin : bot detected

### ISSUE THREAD

**Issue by lasconic** [lasconic/finary_uapi#187]
See https://community.finary.com/t/api-finasync-extension-open-source/10104/20

After a  call to `signin`  we get 

```json
{
    "errors": [
        {
            "message": "Bot traffic was detected. Please try again or contact support.",
            "code": "bot_detected"
        }
    ]
}
```

**Comment by lasconic** [comment]
My first thoughts:

* Could be as simple as a referer or user-agent check ? (I doubt but it's easy to try...)
* Could because there is no delay between the call with username and password and the call with the 2FA code... We could add a delay but it could fail because of the validity period of the 2FA code. Or we change our API, and we ask for the code interactively...

Any other ideas ?

**Comment by lasconic** [comment]
Actually, the error occurs right away on the first code, we don't even get the chance to submit the 2FA code.

**Comment by lasconic** [comment]
The following works, not sure why it doesn't work with requests yet.
```
curl 'https://clerk.finary.com/v1/client/sign_ins' \
  --http1.1 \
  -H 'User-Agent: finary_uapi 0.2.0' \
  -H 'Connection: keep-alive' \
  -H 'Origin: https://app.finary.com' \
  -H 'Referer: https://app.finary.com/' \
  -H 'Accept-Encoding: identity' \
  --data-raw 'password=PASSWORD&identifier=EMAIL' \
  --verbose
```

**Comment by lasconic** [comment]
@nmathey @MadeInPierre
A possible solution is to switch to httpx. There must be a subtle difference between requests and curl or httpx for requests to be detected as a bot with the configuration by default. I can't find this difference... Using httpx, most of the things seems to work for now... Is it a problem for your respective projects to use httpx instead of requests ?

**Comment by lasconic** [comment]
Note that even with httpx running from github actions doesn't work... clerk detects a bot.

**Comment by lasconic** [comment]
See https://github.com/lasconic/finary_uapi/tree/switch_httpx

**Comment by nmathey** [comment]
Moving away from requests shouldn't be a problem for me. 
So from GitHub it still doesn't but you were able to get it working from your wrapper?

**Comment by lasconic** [comment]
Yes, the tests don't work from Github Actions server but they run fine from my home computer. You can try to signin with the main branch (and should get the bot detected error) and with [this branch](https://github.com/lasconic/finary_uapi/tree/switch_httpx) and you should get success.

**Comment by nmathey** [comment]
Got same error
```
INFO:httpx:HTTP Request: POST https://clerk.finary.com/v1/client/sign_ins "HTTP/1.1 403 Forbidden"
{
    "errors": [
        {
            "message": "Bot traffic was detected. Please try again or contact support.",
            "code": "bot_detected"
        }
    ]
}
```

Did your branch include all headers additions you tested with curl?

 ```
--http1.1 \
  -H 'User-Agent: finary_uapi 0.2.0' \
  -H 'Connection: keep-alive' \
  -H 'Origin: https://app.finary.com' \
  -H 'Referer: https://app.finary.com/' \
  -H 'Accept-Encoding: identity' \
```

**Comment by lasconic** [comment]
See https://community.finary.com/t/api-finasync-extension-open-source/10104/24 for the current official status (by Finary CTO).

The behavior described in this bug is "expected". For me, the HTTPX branch works. I tried several headers but of course, I made sure to use exactly the same one while testing. I had the same header in the main branch (using requests), with curl and with HTTPX (including forcing HTTP 1.1 when possible). Clerk does smart things for their bot detection for sure.

For now, I would like to get more data. Does the main branch work for you ? Does the [HTTPX](https://github.com/lasconic/finary_uapi/tree/switch_httpx) branch work ?

**Comment by nmathey** [comment]
Sorry I'm bit confuse, and might be a dummy question, but can you confirm headers you successfully tested are hardcoded in your Httpx branch?
Or should I configured them myself somewhere?

**Comment by lasconic** [comment]
@nmathey they aren't hardcoded but the HTTPX branch works for my laptop as it, without configuring anything.

**Comment by JulesGuesnon** [comment]
Hey! 
I've been developing my own client with [reqwest](https://docs.rs/reqwest/latest/reqwest/) to automatically update balances of manual accounts, and I faced the same problem while developing:
- I tried to put the same headers than the browser but I was detected as a bot
- I tried CURL and it works

I was using this repo as a reference for implementation (thanks for the awesome work btw), and I naively tried to set my user-agent to: `finary_uapi 0.2.0`, and it worked. 

I haven't tried anything else since then, but my guess is that Clerk detects http clients that try to behave like browsers as they're probably bots most of the times, and allows clients that are transparent and saying that they're an API.

So basically, I don't think it's an issue related to the http client, but it's probably linked to headers. Hope those information can help you!

**Comment by clemlesne** [comment]
I just pushed a PR to solve that issue: https://github.com/lasconic/finary_uapi/pull/210.

**Comment by lasconic** [comment]
Thank you for the new look. I checked every possible combinations 6 months ago but your headers seems to just work ! It can't hurt to merge the PR.

**Comment by lasconic** [comment]
@all-contributors please add @clemlesne for code

**Comment by lasconic** [comment]
@all-contributors please add @clemlesne for code

**Comment by nmathey** [comment]
Still getting same message with finary-uapi 0.2.1 b9a9a3f
Is that 100% time working on your side ?

**Comment by OxyFlax** [comment]
Don't you think Finary is actively monitoring our solutions here to update their bot detection with any found solution? I mean if they want to close their api it's not a big work to track this issue and ban any further solution proposed... :(

**Comment by lasconic** [comment]
Let me add that the tests running on Github actions are not working and we still get a "bot traffic detected". See https://github.com/lasconic/finary_uapi/actions/runs/10584142604/job/29327710214#step:8:1424
Maybe Clerk uses an IP blacklist ? No idea.

**Comment by lasconic** [comment]
Yes, they provide testing tokens for development instances : https://clerk.com/docs/testing/overview#testing-tokens so they definitely do something against github actions runners.

**Comment by nmathey** [comment]
I'm using finary_uapi out of the box, and getting the error straight from on sign-in.
I'm using the same IP for both finary_uapi and browsing finary app/portal, so I don't think my IP is blacklisting (?)

**Comment by clemlesne** [comment]
> I'm using finary_uapi out of the box, and getting the error straight from on sign-in.

It works for me, repeatably, for multiple days in a row.

Can you share reproduction steps?

> I'm using the same IP for both finary_uapi and browsing finary app/portal, so I don't think my IP is blacklisting (?)

For sure that's not IP blacklist, at least for your residential IP. Those countermeasures are really rare with SaaS systems (imagine a VPN or a company proxy, you could disrupt hundred of legitimate users).

**Comment by nmathey** [comment]
I'm just trying the signin method (with 2FA or not):

```
abc@d3a3d0a14c9b:~/workspace/finary_uapi$ python -m finary_uapi signin
{
    "errors": [
        {
            "message": "Bot traffic was detected. Please try again or contact support.",
            "code": "bot_detected"
        }
    ]
}
```

**Comment by clemlesne** [comment]
This error should have raised an exception with a user-friendly format (see https://github.com/lasconic/finary_uapi/blob/cf17b614100875156e9052ac3c4864b78c4a5c1f/finary_uapi/signin.py#L63). You're maybe not on the right code-base version, I guess.

**Comment by nmathey** [comment]
I'm not a git expert but my understanding is I'm using latest branch : am I correct ?

```
abc@d3a3d0a14c9b:~/workspace/finary_uapi$ cat .git/refs/heads/${branch-main}
cf17b614100875156e9052ac3c4864b78c4a5c1f
```

**Comment by lasconic** [comment]
@kipavy I doubt they read the issues here. You can add your voice here : https://community.finary.com/t/dev-api-finary/829
And of course, you can ask your data via email (yeah I know...) : dpo@finary.com
We can hope that if enough people do it, they will provide export at least...

**Comment by kipavy** [comment]
> [@kipavy](https://github.com/kipavy) I doubt they read the issues here. You can add your voice here : [community.finary.com/t/dev-api-finary/829](https://community.finary.com/t/dev-api-finary/829) And of course, you can ask your data via email (yeah I know...) : [dpo@finary.com](mailto:dpo@finary.com) We can hope that if enough people do it, they will provide export at least...

yeah sure, btw I've managed to login with my fix https://github.com/kipavy/finary_uapi/commit/7722125f24f4046a436f20a59a2906ca8047d6ff (I just used Claude 4.5 dumbly, didnt want to spend much time on it) (also signin command should be renamed to login imo).

**Comment by MadeInPierre** [comment]
@kipavy Thanks for sharing, I've tried your fix (great idea to use `curl_cffi`!) but getting a "signed out" when signin in...

```bash
jack@jack-framework-13:~/dev/invest/finary_uapi$ python -m finary_uapi signin
{
    "errors": [
        {
            "message": "Signed out",
            "long_message": "You are signed out",
            "code": "signed_out"
        }
    ],
    "clerk_trace_id": "e6e347e3b13936a4e8e29b16cb6fdc14"
}
```

By any chance, have you encountered the same issue?

**Comment by kipavy** [comment]
> [@kipavy](https://github.com/kipavy) Thanks for sharing, I've tried your fix (great idea to use `curl_cffi`!) but getting a "signed out" when signin in...
> 
> jack@jack-framework-13:~/dev/invest/finary_uapi$ python -m finary_uapi signin
> {
>     "errors": [
>         {
>             "message": "Signed out",
>             "long_message": "You are signed out",
>             "code": "signed_out"
>         }
>     ],
>     "clerk_trace_id": "e6e347e3b13936a4e8e29b16cb6fdc14"
> }
> 
> By any chance, have you encountered the same issue?

I think I got the same, didnt understand either...

**Comment by lasconic** [comment]
main branch now uses curl-cffi. It works for me, at least for "GET" calls. Many things changed in the API, so I removed the deleted endpoints. There are many new endpoints used by the webapp that we could implement.
